### PR TITLE
Migrate to matrix-nio

### LINKED ISSUE #542: Matrix client dependency
When I look at the github page for the dependency of matrix_client at https://github.com/matrix-org/matrix-python-sdk it reads:

> We strongly recommend using the matrix-nio library rather than this sdk. It is both more featureful and more actively maintained.

Is there a plan to change to the matrix-nio library as a client dependency?

### PR DESCRIPTION
**Pull request by doubleloop**
Fixes #542

### CODE REVIEW

**Review by thp — CHANGES_REQUESTED** [review]
This probably needs a section in the "Deprecated features" documentation on how to migrate from an old matrix configuration to the new one, otherwise people will have their matrix notifications broken (either silently or loudly), which isn't nice. Is there a way we can auto-migrate from the old way to the new way, so Matrix notifications just keep running? Alternatively, should this be a "matrix-nio" reporter and we leave support for the old one in place, but the documentation will just mention that a matrix reporter existed, but new installations should use "matrix-nio"?

**Comment by thp** [comment]
Any news on this? Do you think you want to work on getting this ready to be merged?

**Comment by doubleloop** [comment]
>needs a section in the "Deprecated features" documentation on how to migrate from an old matrix configuration to the new one, otherwise people will have their matrix notifications broken (either silently or loudly), which isn't nice.

I added special case so that if configuration compatible with `matrix_client` is found, it will still work as expected (with no encryption support). I also added `Deprecated Features` section describing changes required for encryption support.

>Do you think you want to work on getting this ready to be merged?
Yes, waiting for some review/feedback and list of things needed to make this ready to be merged.

File: docs/source/deprecated.rst
Comment (thp) [inline docs/source/deprecated.rst:12]: "support encrypted rooms" -> "supports encrypted rooms"

File: docs/source/deprecated.rst
Comment (thp) [inline docs/source/deprecated.rst:47]: Just replace this with a link to the Matrix reporter documentation (merging anything that's here and not there to there).

File: docs/source/reporters.rst
Comment (thp) [inline docs/source/reporters.rst:234]: "to acquire room ID" -> "to acquire the room ID"

File: docs/source/reporters.rst
Comment (thp) [inline docs/source/reporters.rst:262]: "approprite" -> "appropriate"

Comment (thp) [reply docs/source/reporters.rst:262]: Remove "(e.g. ``~/.local/share/urlwatch/`` on linux)" and replace with something like "(relative to urlwatch's configuration directory)".

File: lib/urlwatch/tests/test_handler.py
Comment (thp) [inline lib/urlwatch/tests/test_handler.py:93]: Should this use a temporary folder instead of the path where test_handler is? If it's installed, it might not be writable to a user wanting to run the tests.

File: docs/source/deprecated.rst
Comment (thp) [inline docs/source/deprecated.rst:19]: Should we warn when loading a config that still has `access_token` set?

File: lib/urlwatch/reporters.py
Comment (thp) [inline lib/urlwatch/reporters.py:1029]: I know this log message has been like this before, but since we touch it, maybe we can just make it something like:

    markdown2 could not be imported, skipping Markdown formatting

File: lib/urlwatch/handler.py
Comment (thp) [inline lib/urlwatch/handler.py:179]: Is this change necessary?

Comment (thp) [reply lib/urlwatch/handler.py:179]: Or in other words, could we just set the data directory as additional parameter, not expose the whole config object?

**Review by thp — CHANGES_REQUESTED** [review]
See comments. We're getting there :)

File: docs/source/deprecated.rst
Comment (doubleloop) [reply docs/source/deprecated.rst:19]: https://github.com/thp/urlwatch/pull/620/files#diff-5ee1abebe5480db745909e862aaa226d8f66babc4ab506b526c49cf7c0059fe0R1034-R1038

File: docs/source/deprecated.rst
Comment (doubleloop) [reply docs/source/deprecated.rst]: could you verify if I have done it correctly?

File: lib/urlwatch/tests/test_handler.py
Comment (doubleloop) [reply lib/urlwatch/tests/test_handler.py]: Made temporary folder. Not sure why `os.path.dirname(__file__)` is used here.
note: what do you thing about renaming `urlwatch_dir` to `urlwatch_config_dir` ?

File: lib/urlwatch/handler.py
Comment (doubleloop) [reply lib/urlwatch/handler.py]: Not sure why not to expose the whole config (other features might require access to command line args or other paths like cache), but I will change this.

> could we just set the data directory as additional parameter
Note that I only renamed existing parameter and did not introduce any new one (`urlwatch_config` -> `urlwatch_obj`). Previous name was misleading (https://github.com/thp/urlwatch/blob/master/lib/urlwatch/main.py#L54)

**Comment by doubleloop** [comment]
I rebased on master and added `review fix` commits (I plan to squash them before merge).

File: lib/urlwatch/reporters.py
Comment (thp) [inline lib/urlwatch/reporters.py:985]: This should most likely be a separate mode from the command line, as we don't want random invocations to be interactive (urlwatch is most likely running as a cronjob or via some other non-interactive means).

Comment (thp) [reply lib/urlwatch/reporters.py:985]: See the `--xmpp-login` command-line option for an example. Maybe combine it like `--test-filter` so that you can do e.g. `--login-reporter xmpp` and `--login-reporter matrix` (as `--xmpp-login` is an interactive thing, maybe just update docs + mention it in the list of deprecated things that `--xmpp-login is now `--login-reporter xmpp`).

File: lib/urlwatch/reporters.py
Comment (thp) [inline lib/urlwatch/reporters.py:1055]: `30000` should probably be an named constant (`MATRIX_CLIENT_SYNC_TIMEOUT_MS`?)

File: docs/source/reporters.rst
Comment (thp) [inline docs/source/reporters.rst:237]: Is there some command-line matrix client that can be used for this without relying on `riot.im` or a similar service? I know that this was in here before, but since one of the advantages of this change is that rooms are now encrypted, the encryption would be kinda pointless if initially we rely on trusting `riot.im`.

**Review by thp — CHANGES_REQUESTED** [review]
See comments. We're getting there!

File: docs/source/reporters.rst
Comment (doubleloop) [reply docs/source/reporters.rst:237]: > Is there some command-line matrix client that can be used for this without relying on riot.im or a similar service?

According to spec, room `name` is not unique, room `alias` is optional so `room ID` in general is the only reliable room identifier (https://matrix.org/docs/spec/client_server/latest#m-room-name). `room ID` is obtained during room creation and stored by the client. Any client used to create room should be able to display `room ID` (I can note in docs that ~riot.im~ element is just an example)

>the encryption would be kinda pointless if initially we rely on trusting riot.im

Why?

**Comment by thp** [comment]
@doubleloop Any updates on this PR?

**Comment by thp** [comment]
Closing for now. Please reopen once rebased.

**Pull request closed without merging**
